R/utils.R

```diff
@@ -0,0 +1,146 @@
+# operations (i.e. those since either the beginning of time, or the most recent
+# `write` operation) to be undone, AND for any future `writeWS` calls to be
+# ignored. A call to `write` will be respected, and will restore normal
+# behavior.
+WSTextWriter <- R6Class("WSTextWriter",
+  private = list(
+    writer = "TextWriter"
+  ),
+  public = list(
+    initialize = function() {
+      private$writer <- TextWriter$new()
+    },
+    close = function() {
+      private$writer$close()
+    },
+    write = function(text) {
+      # If an error is going to happen while evaluating `text`, let's make it
+      # happen before we mutate any of our internal state
+      force(text)
+
+      # Stop eating whitespace
+      private$writer$setSuppress(FALSE)
+
+      private$writer$write(text)
+
+      # Set a bookmark here, so that future calls to eatWS() don't affect this
+      # write
+      private$writer$savePosition()
+    },
+    # Write whitespace. If suppressWhitespace() was called and
+    # its effect has not been canceled, then this method no-ops.
+    # @param text Single element character vector containing only
+    #   whitespace characters
+    writeWS = function(text) {
+      private$writer$write(text)
+    },
+    readAll = function() {
+      private$writer$readAll()
+    },
+    # Removes both recent and upcoming whitespace writes
+    eatWS = function() {
+      # Undo recent whitespace writes
+      private$writer$restorePosition()
+      # Ignore upcoming whitespace writes
+      private$writer$setSuppress(TRUE)
+    }
+  )
+)
+
+#' TextWriter class
+#'
+#' A class that manages the gradual concatenation of text. It
+#' provides three additional features (over a normal connection)
+#' that are important to us for tag writing:
+#'
+#' 1. Text is automatically converted to UTF-8.
+#'
+#' 2. Ability to bookmark (save) a write position, and jump back
+#'    to it (restore). Only one bookmark at a time is allowed.
+#'    Restoring a bookmark essentially truncates content that
+#'    was written after the bookmark was set.
+#'
+#' 3. Ability to suppress writes (setSuppress); while suppression
+#'    is set to true, all write operations no-op.
+#'
+#' @noRd
+#' @importFrom R6 R6Class
+TextWriter <- R6Class("TextWriter",
+  private = list(
+    con = "ANY",
+    marked = numeric(1),
+    suppressing = logical(1)
+  ),
+  public = list(
+    initialize = function() {
+      # We use a file() here instead of textConnection() or paste/c to
+      # avoid the overhead of copying, which is huge for moderately
+      # large numbers of calls to .$write(). Generally when you want
+      # to incrementally build up a long string out of immutable ones,
+      # you want to use a mutable/growable string buffer of some kind;
+      # since R doesn't have something like that (that I know of),
+      # file() is the next best thing.
+      private$con <- file("", "w+b", encoding = "UTF-8")
```

Comment:
I can't tell from the `file` docs whether this is in-memory or on disk. It's an unlinked temporary file? I'm not sure what the behavior is there.

I ask because if this does live in memory I wonder if it would be worth having a configurable maximum buffer size.

Reply:
I guess either way it's going to get read into memory during `readAll`, so maybe the `file` behavior is moot.